src/main.cpp

```diff
@@ -6008,10 +6008,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                             pindexLast->GetBlockHash().ToString(), pindexLast->nHeight);
                 }
                 if (vGetData.size() > 0) {
-                    if (nodestate->fSupportsDesiredCmpctVersion && vGetData.size() == 1 && mapBlocksInFlight.size() == 1 && pindexLast->pprev->IsValid(BLOCK_VALID_CHAIN)) {
+                    if (nodestate->fSupportsDesiredCmpctVersion && vGetData.size() == 1 && mapBlocksInFlight.size() == 1 && pindexLast->pprev->IsValid(BLOCK_VALID_CHAIN) && mempool.size() > 1) {
                         // We seem to be rather well-synced, so it appears pfrom was the first to provide us
-                        // with this block! Let's get them to announce using compact blocks in the future.
-                        MaybeSetPeerAsAnnouncingHeaderAndIDs(nodestate, pfrom, connman);
+                        // with this block! Let's get them to announce using compact blocks in the future, unless
+                        // either of us are blocks-only and would prefer to save bandwidth.
+                        if (pfrom->fRelayTxes && !GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY)) {
```

Comment:
The thinking there was that if the peer is trying to economize bandwidth, we shouldn't ask them to send the unsolicited compact blocks (HB mode). We'll still request compact blocks from them when advertised.  I don't fee super strongly, if they're trying to economize on bandwidth they should probably delay their relays artificially.  We don't have a NAK for HB mode requests though.